s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala

```diff
@@ -0,0 +1,197 @@
+package akka.stream.alpakka.s3.impl
+
+import java.nio.file.Paths
+import java.time.LocalDate
+
+import akka.NotUsed
+import akka.actor.ActorSystem
+import akka.event.Logging
+import akka.http.scaladsl.Http
+import akka.http.scaladsl.model.{ HttpRequest, HttpResponse, ResponseEntity, Uri }
+import akka.http.scaladsl.unmarshalling.{ Unmarshal, Unmarshaller }
+import akka.stream.{ Attributes, Materializer }
+import akka.stream.alpakka.s3.{ MemoryBufferType, DiskBufferType, S3Settings }
+import akka.stream.alpakka.s3.auth.{ AWSCredentials, CredentialScope, Signer, SigningKey }
+import akka.stream.scaladsl.{ Flow, Keep, Sink, Source }
+import akka.util.ByteString
+
+import scala.collection.immutable.Seq
+import scala.concurrent.{ ExecutionContext, Future }
+import scala.util.{ Failure, Success }
+
+final case class S3Location(bucket: String, key: String)
+
+final case class MultipartUpload(s3Location: S3Location, uploadId: String)
+
+sealed trait UploadPartResponse {
+  def multipartUpload: MultipartUpload
+  def index: Int
+}
+
+final case class SuccessfulUploadPart(multipartUpload: MultipartUpload, index: Int, etag: String) extends UploadPartResponse
+final case class FailedUploadPart(multipartUpload: MultipartUpload, index: Int, exception: Throwable) extends UploadPartResponse
+
+final case class FailedUpload(reasons: Seq[Throwable]) extends Exception
+final case class CompleteMultipartUploadResult(location: Uri, bucket: String, key: String, etag: String)
+
+private[alpakka] final class S3Stream(credentials: AWSCredentials, region: String, val settings: S3Settings)(implicit system: ActorSystem, mat: Materializer) {
+  import Marshalling._
+
+  val MinChunkSize = 5242880
+  val signingKey = SigningKey(credentials, CredentialScope(LocalDate.now(), region, "s3"))
+
+  def this(credentials: AWSCredentials, region: String = "us-east-1")(implicit system: ActorSystem, mat: Materializer) =
+    this(credentials, region, S3Settings(system))
+
+  def download(s3Location: S3Location): Source[ByteString, NotUsed] = {
+    import mat.executionContext
+    Source.fromFuture(signAndGet(HttpRequests.getRequest(s3Location)).map(_.dataBytes))
+      .flatMapConcat(identity)
+  }
+
+  /**
+   * Uploads a stream of ByteStrings to a specified location as a multipart upload.
+   *
+   * @param s3Location
+   * @param chunkSize
+   * @param chunkingParallelism
+   * @return
+   */
+  def multipartUpload(s3Location: S3Location, chunkSize: Int = MinChunkSize, chunkingParallelism: Int = 4): Sink[ByteString, Future[CompleteMultipartUploadResult]] = {
+    val flow = chunkAndRequest(s3Location, chunkSize)(chunkingParallelism)
+    (if (settings.debugLogging)
```

Comment:
I do not think we need this. `flow.log` below logs at the debug level. And if the user has `akka.loglevel != DEBUG` then these log messages will be dropped.